### DISCUSSION TITLE
`deps` will force rerender of the whole form?

### DISCUSSION THREAD

**Discussion by tkharuk** [react-hook-form/react-hook-form#7831] [Q&A] [answered]
Hi,
I have an array of arrays, with, possibly a huge amount of fields so performance is pretty important.

After validation got triggered, the input with registered `deps` will cause renders of **all inputs** in the form.

https://user-images.githubusercontent.com/1205052/154084476-65fb34fa-23fe-4f66-9205-b160cdfca4b0.mov

#### Notes:
- The fields are implemented with `Controller`
- `deps` value in video is something like `['rules.0.nightRanges.0.from']`

**Comment by Moshyfawn** [comment]
I can't remember where I've seen the info, so **take it with a grain of salt**, but I think there was some discussion where it was said that trigger is supposed to `trigger` the re-render of the whole form _(it's the underling API method that uses the `deps` prop internally)_.. But now, that I found #6899, I'm not sure.. @bluebill1049 can you confirm / deny, please?

**Reply by bluebill1049** [reply]
it depends on your `formState` subscription if you subscribe at the root. yes, it would re-render the whole form. also for array-like syntax, there is no optimization targeted on field level, pretty much the same as trigger() and pick the result.

**Reply by tkharuk** [reply]
The only state I use is the `fieldState` from `useController`.
And those are actually the things that gets triggered.
In video, not the whole page is rerendered but just every field.

So, it is expected of `deps` to cause this?
Are there any ways for me to make the scope of impact smaller?

**Reply by bluebill1049** [reply]
yes, it's expected. 

https://react-hook-form.com/api/useform/trigger

> Isolate render optimisation only applicable for targeting a single field name with string as payload, when supplied with array and undefined to trigger will re-render the entire formState.

**Reply by bluebill1049** [reply]
The reason why it's limited to `string` is that for string it's limited and precise to single node validation and it's guaranteed. However, for array (not tuple), it's not a fixed amount, and doing optimization on that level could be doing the opposite, so we decided to for `array` or `undefined` to flush down the entire formstate, but single (string) field to do precise value/formState update.

**Reply by tkharuk** [reply]
@bluebill1049 in documentation and typings, `deps` is described as `string[]`, thus I haven't even tried `string`:

<p align="center">
<img width="250" alt="Screenshot 2022-02-23 at 4 22 49 PM" src="https://user-images.githubusercontent.com/1205052/155338008-5de14dfc-8598-4538-8880-8b7ae710758a.png">
<img width="400" alt="Screenshot 2022-02-23 at 4 16 22 PM" src="https://user-images.githubusercontent.com/1205052/155338036-2ef15507-cbfe-4bb3-ae84-8077981a7b6a.png">
</p>
 





---
Just `string` gives a much better result for my particular case 👍🏼

https://user-images.githubusercontent.com/1205052/155336763-6c1cf492-45d6-47b2-b67b-e52d1bcf008f.mov

**Reply by bluebill1049** [reply]
https://github.com/react-hook-form/react-hook-form/pull/7901

**Comment by tkharuk** [comment]
From some observations:
- `{watch} = useFormContext();` - render everywhere
- `useWatch` - render only on the field array level
- `setError` - render only on the field level
- `clearErrors` - render everywhere, can use `{invalid} = getFieldState` to not call excessively

Best I could achieve so far:

https://user-images.githubusercontent.com/1205052/154274600-180433e7-01ce-469d-b765-d3f752538620.mov

Still would need to align "mode" and "revalidateMode".


Really didn't expect `watch` and `useWatch` to behave differently.
Would need to have some "best practices" and conventions in team, which I hate :D


Do you think this is something that can/should be improved on lib level?
Given the fact one can improve it with existing methods, it should be pretty safe, right?

**Comment by bluebill1049** [comment, marked as answer]
> Really didn't expect watch and useWatch to behave differently.

please refer to the doc, https://react-hook-form.com/api/useform/watch there is a useful video there as well.

> Do you think this is something that can/should be improved on lib level?

we couldn't, because watch flush update at the `useForm` level, while `useWatch` will be at its hook level.
